src/tools/wasm-ctor-eval.cpp

```diff
@@ -421,11 +421,149 @@ struct CtorEvalExternalInterface : EvallingModuleRunner::ExternalInterface {
+  // code, we end up with some live allocations in the interpreter, which we
+  // need to somehow serialize into the wasm module. We will put each such live
+  // GC data item into its own "defining global", a global whose purpose is to
+  // create and store that data. Each such global is immutable, and has the
+  // exact type of the data, for simplicity. Every other reference to that GC
+  // data in the interpreter's memory can then be serialized by simply emitting
+  // a global.get of that defining global.
   void applyGlobalsToModule() {
     Builder builder(*wasm);
-    for (const auto& [name, value] : instance->globals) {
-      wasm->getGlobal(name)->init = builder.makeConstantExpression(value);
+
+    if (!wasm->features.hasGC()) {
+      // Without GC, we can simply serialize the globals in place as they are.
+      for (const auto& [name, values] : instance->globals) {
+        wasm->getGlobal(name)->init = getSerialization(values);
+      }
+      return;
     }
+
+    // We need to emit the "defining globals" of GC data before the existing
+    // globals, as the normal ones may refer to them. We do this by removing all
+    // the existing globals, and then adding them one by one, during which time
+    // we call getSerialization() for their init expressions. If their init
+    // refes to GC data, then we will allocate a defining global for that data,
+    // and refer to it. Put another way, we place the existing globals back into
+    // the module one at a time, adding their dependencies as we go.
+    auto oldGlobals = std::move(wasm->globals);
+    wasm->updateMaps();
+
+    // The process of allocating "defining globals" begins here. Clear any
+    // previous state.
+    definingGlobals.clear();
+
+    for (auto& oldGlobal : oldGlobals) {
+      // Serialize the global. While doing so pass in the name of this global,
+      // as it may be used as the defining global for it, if one does not
+      // already exist (this optimization prevents allocating an extra global
+      // just to global.get its value right after it).
+      Name name;
+      if (!oldGlobal->mutable_ && oldGlobal->type == oldGlobal->init->type) {
+        // This has the properties we need of a defining global - immutable and
+        // of the precise type - so use it.
+        name = oldGlobal->name;
+      }
+
+      // If there is no interpreter value for this global, that means the
+      // interpreter never executed code to create it. That happens when this is
+      // a new global that we have added after running the interpreter, which
+      // means this is one of the new defining globals, which we can simply
+      // skip here - if such a global is still needed, it will show up as a
+      // dependency of something, and be emitted at the right time and place.
+      auto values = instance->globals[oldGlobal->name];
+      if (values.empty()) {
+        continue;
+      }
```

Comment:
Interesting question. We can definitely avoid the empty values by using an iterator to check if there is anything there at all... not sure why I didn't do that before. Added now.

But maybe that doesn't address your entire concern? We still have this special code here. It seems hard to avoid it. Perhaps we could maintain a separate list of "new defining globals declared, that have no values". But that would add complexity, I think.